### DISCUSSION TITLE
Export as BMP bitmap

### DISCUSSION THREAD

**Discussion by Mat-thieu** [samizdatco/skia-canvas#64] [Q&A]
Hi,

I have a usecase where we'd like to export our canvas to BMP, so just export the canvas as-is without the need of rasterizing or transforming it to vector graphics.

Is this a possible output option at the moment? If not, would it be possible to add?

**Comment by samizdatco** [comment]
The underlying rust library supports _decoding_ of [BMP files](https://github.com/rust-skia/rust-skia/tree/master/skia-safe#codecs) but can't actually generate them I'm afraid.

One possible workaround would be to use [`getImageData`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData) to retrieve the raw pixel values and then pass them to a library like [bmp-js](https://www.npmjs.com/package/bmp-js) to do the actual encoding. It looks like the [Jimp](https://www.npmjs.com/package/jimp) library also supports BMP conversions.

**Comment by Mat-thieu** [comment]
Ah great, thank you, I will look into these. I think getImageData might already contain what I'm looking for!

**Comment by aastefian** [comment]
Hello hello, I stumbled upon this issue, and I gave `getImageData()` a try, it works, but it is very very slow (it can take 1.5s to extract buffer from a canvas of 1920 x 1080), and in my use case I am generating a video from the canvas data which means I generate 25 frames(canvas getImageData() calls) per second. That is at least 25 * 1.5s per second of video generated (excluding encoding and other stuff)

Are there any plans or considerations of exposing the image buffer of the canvas directly ? Or, do you have any ides why that function might be so slow ?

 Something like a simple array of rgba ordered pixels: [R,  G, B, A, R, G, B, A.....]. That would be generic enough to do anything with it.

**Reply by ggolda** [reply]
Having the same issue, have you found any workaround for that of how to extract pixel buffer?

**Reply by aastefian** [reply]
hello, in the end i used the same method and it turned out the slowness for me was coming from the fact that i was using the same instance of canvas and skia-canvas had an issue that after many drawImage calls it woudl get really slow, so i destroy and recreate the canvas for every frame i generate which is slower in theory but in this case makes it faster because it avoids the bug.

Not sure if this is an issue in the latest version of skia-canvas still though.

This is the snippet i use, keep in mind:
- alpha is not premultiplied - after I extracted the buffer, i had to apply alpha to the channels by applying an algorithm i found online, you might not have to do that, i had to do it because the software im using after this needed premultiplied alpha
- i had to convert the array from uint8clamped to uint8 when using streams in nodejs
![image](https://github.com/samizdatco/skia-canvas/assets/15310971/60962745-9611-439b-860c-2efd747438c1)

**Reply by mpaperno** [reply]
Hi,

I recently added raw pixel export option for the whole canvas [in my fork](https://github.com/mpaperno/skia-canvas/blob/mp/main/CHANGELOG.md).  It's much faster in relative terms than `getImageData()` because it's async and uses an actual thread for processing, just like `toBuffer()`, et al.  There is an option to premulitply alpha channel as well.

Perhaps worth a try if a better solution hasn't presented itself.
